xapian-letor/bin/xapian-prepare-trainingfile.cc

```diff
@@ -74,7 +75,11 @@ try {
 		have_database = true;
 		break;
 	    case 'm':
-		msize = atoi(optarg);
+		unsigned int size;
+		if (!parse_unsigned(optarg, size)) {
+		    throw "Mset size must be >= 0";
+		}
+		msize = size;
```

Comment:
Currently `msize` is of type `int` [*], so considering when `int` is 32 bits, specifying `-m3000000000`, size is `3000000000` and `msize` is `-1294967296`.

[*] This seems wrong to me - it ought to be `Xapian::doccount`, since that's the what the `msetsize` parameter `prepare_training_file()` takes is, and that's the appropriate type for a count of Xapian documents.  If we change `msize` to be `Xapian::doccount` then we can just pass that to `parse_unsigned()`, eliminate `size` and this problem goes away.

I'm not sure throwing a `const char*` error message is a great approach here - it reduces flexibility since the error message really needs to be a constant (or else you have to mess around with allocating buffers or static `std::string`).  It seems simpler to just report the error and exit in place of the `throw`, and that's what we generally do in similar situations.  It's not like our caller might want to handle the exception in an alternative way, because we're parsing command line options in `main()` here.

Also the error message should really go to `cerr` not `cout` (the `--help` output isn't an error, so that rightly goes to `cout`).

Reply:
fixed in a fixup commit.